### ISSUE TITLE
Training Set Size

### ISSUE THREAD

**Issue by chaitanyamalaviya** [UniversalDependencies/UD_Ukrainian-IU#4]
Hi, 

I found it strange that the training set for Ukrainian has only 37 sentences in the latest commit, while the dev set has a much larger number of sentences (826). Is this something done on purpose?

**Comment by jnivre** [comment]
This is a consequence of the CoNLL shared task, where all test sets had to contain at least 10,000 words. Of the remaining data, another 10,000 words were reserved for the dev set. Whatever remained after that became the training set. For some languages, this left very little data for training (and we then allowed participants to use the dev set for training as well). The rationale is that all new data will in the future be added to the training set, so that this set can keep growing while the dev and test sets remain stable.

**Comment by msklvsk** [comment]
Four times more data coming soon! However, we won't be able to maintain style and complexity balance if we only add to train. @jnivre Can we break set compatibility in 2.1 and re-split? Plus there are so many fixes and improvements, 2.0 models can be considered “broken” anyway.

**Comment by jnivre** [comment]
Resplitting is tricky, and we definitely don't want to move data from train to test or vice versa, because of the risk that people get test data in their training set if they mix versions. (Even if you know that 2.0 models are broken, you cannot guarantee that other people do not make foolish things.) We could possibly add data to all sets, though. I guess we need some guidelines on this.

**Comment by msklvsk** [comment]
We were on a very tight schedule to annotate first 20K for the shared task, so we used morphological annotation we already had, which was mostly fiction and other belles-lettres. Now we have diluted it up to 100K with news and articles, but dev/test is already full. Without re-splitting you will practically train on non-fiction, but test on fiction. Users who understand the concept of versioning would get suboptimal models for the sake of users who mix versions (those would also get suboptimal anyway, nobody would be able to build a proper model). I do agree data should not be moved around the sets for no reason, but 2.0 was an early version for UD-Uk, and the split was dictated by the shared task. We can guarantee dataset stability since 2.1.

**Comment by jnivre** [comment]
I see your point of course, but you also have to appreciate the fact that, with nearly 100 treebanks, both developers and users need to be able to rely on general rules that apply to all treebanks, without having to keep track of individual exceptions. Otherwise, the whole project will collapse sooner or later. Hopefully, we can come up with a solution to your problem, but it needs to a general solution, not an ad hoc exception. I would therefore like to invite @dan-zeman and @fginter to the discussion, since they are the ones that have to manage all releases.

**Comment by dan-zeman** [comment]
My opinion is that _"don't move sentences between test and training"_ is a good general rule for data providers but not something we should emphasize and guarantee for the users. (We actually do not guarantee it because we do not test it automatically and we do not report treebanks as invalid if comparing versions reveals such a change.)

I think exceptions to the general rule should be possible when there are good reasons for the exception, it has been discussed with us and the data providers promise to plan ahead so that the same discussion about the same treebank does not take place before every release. (BTW, the CoNLL shared task restriction was somewhat poor planning on the side of us, the task organizers.)

I think that the primary responsibility for not doing foolish things should rest with the users of the data. (And we cannot prevent them from doing other foolish things even if our data never change.)

**Comment by arademaker** [comment]
At some point I would also like to consider re-splitting of sentences of UD_Portuguese treebank. The reason is that sentences were shuffled before the split in the current distribution and we believe keeping the original text flow would allow other tasks such as co-references etc

**Comment by jnivre** [comment]
Maybe this is a separate discussion, but we also need to provide general guidelines for new (and growing= treebanks. If someone only has 2000 words in the first release, should they only release a test set? And then keep adding to it until they have 10 000 words? After that, should they first provide a dev set or a training set?

**Comment by dan-zeman** [comment]
Agreed that test set goes first. It is an improvement that we now can have treebanks consisting of test data only. Previously, if you had just 200 sentences, you were still required to split it to, say, 150 training, 25 dev and 25 test. Although the best thing to do when any "training" is involved, would be to split it 10-fold and cross-validate.

And if, couple of releases later, you have 13000 tokens, I would not vote for splitting to 10000 test and 3000 training because I find it weird to have training sets smaller than test sets. Instead, I would keep growing the test set until at least 20000 tokens, then allow a re-split to training and test. And probably something similar with dev. I do not know what exactly, but for me the dev set is the least important part of the three.

**Comment by jnivre** [comment]
But we don't have any guidelines for this under "how to contribute", do we?

**Comment by dan-zeman** [comment]
PS: the above suggestion does not work well with a possible future shared tasks where we have to separate test data from training. So if such treebanks are to be included in a task, the real test data will have to be cut and set aside at least for the purpose of the task.

**Comment by dan-zeman** [comment]
http://universaldependencies.org/release_checklist.html#repository-and-files
has some guidelines but they are obsolete (focused on UD 2.0 and the 2017 shared task).

**Comment by msklvsk** [comment]
Or, there can be multiple standoff splits for different purposes and a single `.conllu` file.
(And a tool that generates multiple `.conllu` files for a specific split.)
(Or a default split plus the ability to generate others.)
(wild idea)

**Comment by amir-zeldes** [comment]
+1 for regulating train-test-dev for growing treebanks, and ideally not shuffling sentences.

We have just over 10K tokens for Coptic and reaching 20 will be a while, but we'll definitely want to re-split too, since in some genres we split a single large document between sets to stratify. In later versions we'll want full coherent texts in each set (also for coref and other discourse tasks)

**Comment by msklvsk** [comment]
The data was eventually resplit in v2.1. This discussion triggered creation of [Data split](http://universaldependencies.org/release_checklist.html#data-split) guidelines. Thanks to @chaitanyamalaviya for bringing it up.
